packages/@aws-cdk/aws-s3-deployment/lib/render-data.ts

```diff
@@ -0,0 +1,81 @@
+import { Stack } from '@aws-cdk/core';
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from '@aws-cdk/core';
+
+export interface Content {
+  readonly text: string;
+  readonly markers: Record<string, any>;
+}
+
+/**
+ * Renders the given string data as deployable content with markers substituted
+ * for all "Ref" and "Fn::GetAtt" objects.
+ *
+ * @param scope Construct scope
+ * @param data The input data
+ * @returns The markered text (`text`) and a map that maps marker names to their
+ * values (`markers`).
+ */
+export function renderData(scope: Construct, data: string): Content {
+  const obj = Stack.of(scope).resolve(data);
+  if (typeof(obj) === 'string') {
+    return { text: obj, markers: {} };
+  }
+
+  if (typeof(obj) !== 'object') {
+    throw new Error(`Unexpected: after resolve() data must either be a string or a CloudFormation intrinsic. Got: ${JSON.stringify(obj)}`);
+  }
+
+  let markerIndex = 0;
+  const markers: Record<string, FnJoinPart> = {};
+  const result = new Array<string>();
+  const fnJoin: FnJoin | undefined = obj['Fn::Join'];
```

Comment:
@rix0rrr how horrible is this hack? Any idea for a less leaky way to do this?

Reply:
I am going to leave it like this, because `Stack.resolve()` doesn't expose the capability to pass in a resolver and I prefer not to touch it right now.